### PR TITLE
Add TypeRefCreator trait

### PR DESCRIPTION
**Pull request by tokatoka**

### CODE REVIEW

**Comment by tokatoka** [comment]
like this? @addisoncrump

**Comment by tokatoka** [comment]
i didn't implement it for all the struct but just implemented for what is needed now

**Comment by domenukk** [comment]
I would do a `get(&ref)` and have the ref take the name as well.
Also, this again isn't a `Has`, it should just be `impl TypeRef` I think

**Comment by addisoncrump** [comment]
Can I modify this @tokatoka? There's a nicer way to do this that requires fewer modifications

**Comment by tokatoka** [comment]
yes

**Comment by addisoncrump** [comment]
Now it should just automagically work for all types that get entered into tuples and searched by `match_name`.

**Comment by tokatoka** [comment]
great 👍

File: fuzzers/forkserver_simple/harness
[the remark this answers is not in the document]
Comment (tokatoka) [reply fuzzers/forkserver_simple/harness:1]: deleted

[the remark this answers is not in the document]
Comment (addisoncrump) [reply fuzzers/forkserver_simple/harness:1]: :joy: Just messing with you. We should probably gitignore these.

[the remark this answers is not in the document]
Comment (tokatoka) [reply fuzzers/forkserver_simple/harness:1]: but how do we check if it's executable

[the remark this answers is not in the document]
Comment (addisoncrump) [reply fuzzers/forkserver_simple/harness:1]: No, I mean these exact file names. You just did `cargo make` or something, right?

**Comment by domenukk** [comment]
How about a `Referencable` trait? It gives the struct the trait that it can be referenced

**Comment by domenukk** [comment]
Also, as mentioned before, I think the ref should directly contain the name, So a real `Reference` instead of a `TypeRef`

**Comment by tokatoka** [comment]
@addisoncrump 
can you do the CoW stuff?
I don't know what your desired change is

**Comment by addisoncrump** [comment]
Sure thing.

**Comment by tokatoka** [comment]
this is ready?

File: libafl/src/stages/tmin.rs [hunk 1]
```diff
@@ -349,33 +349,20 @@ where
...
 pub struct MapEqualityFeedback<M, S> {
-    name: String,
-    obs_name: String,
+    name: Cow<'static, str>,
+    obs_name: Cow<'static, str>,
     orig_hash: u64,
     phantom: PhantomData<(M, S)>,
 }
 
-impl<M, S> MapEqualityFeedback<M, S> {
```
Comment (tokatoka) [inline libafl/src/stages/tmin.rs:358]: this guy got deleted for mistake?

Comment (addisoncrump) [reply libafl/src/stages/tmin.rs:358]: Nope, not accidental -- this one should not be publicly exposed.

**Comment by addisoncrump** [comment]
@tokatoka It's not ready, no, I'm working through CI fixes.

File: libafl_bolts/src/tuples.rs [hunk 2]
```diff
@@ -524,6 +526,47 @@ where
     }
 }
 
+/// Structs that has `TypeRef`
+/// You should use this when you want to avoid specifying types using `match_name_type_mut`
+pub trait TypeRefCreator {
```
Comment (domenukk) [inline libafl_bolts/src/tuples.rs:531]: `Referenceable`?

**Comment by tokatoka** [comment]
can you finish this @addisoncrump 
or i shoud just fix them?

**Comment by tokatoka** [comment]
then alloc shouldn't be a feature anymore?
because what can we do without a name now??

**Comment by tokatoka** [comment]
@domenukk 
> I think the ref should directly contain the name, So a real Reference instead of a TypeRef

This makes it hard to use.

For example if we do like this
```
#[derive(Debug, Clone, Copy)]
#[cfg(feature = "alloc")]
pub struct TypeRef<'a, T: ?Sized> {
    name: &'a Cow<'static, str>,
    phantom: PhantomData<T>,
}
```

Then this code from forkserver_libafl_cc won't compile
```
    let observer_ref = edges_observer.type_ref();

    let mut tokens = Tokens::new();
    let mut executor = ForkserverExecutor::builder()
        .program(opt.executable)
        .debug_child(debug_child)
        .shmem_provider(&mut shmem_provider)
        .autotokens(&mut tokens)
        .parse_afl_cmdline(args)
        .coverage_map_size(MAP_SIZE)
        .timeout(Duration::from_millis(opt.timeout))
        .kill_signal(opt.signal)
        .build(tuple_list!(time_observer, edges_observer))
        .unwrap();

    if let Some(dynamic_map_size) = executor.coverage_map_size() {
        executor
            .observers_mut()
            .match_by_ref_mut(observer_ref)
            .unwrap()
            .as_mut()
            .truncate(dynamic_map_size);
    }
```
because edges_observer is moved to the `executor` then we can't use observer_ref later in `match_by_ref_mut()`

**Comment by addisoncrump** [comment]
@tokatoka in that case, you can clone the Cow and only if necessary will it create a memory reference. It can be safely held without reference to the original observer.

```rs
#[derive(Debug, Clone)]
#[cfg(feature = "alloc")]
pub struct TypeRef<'a, T: ?Sized> {
    name: Cow<'static, str>,
    phantom: PhantomData<T>,
}
```

**Comment by addisoncrump** [comment]
And yes, please finish this without me, I am currently rushing a paper submission :face_with_spiral_eyes:

File: libafl_bolts/src/tuples.rs [hunk 3]
```diff
@@ -524,6 +538,54 @@ where
     }
 }
 
+/// Structs that has `ReReference `
+/// You should use this when you want to avoid specifying types using `match_name_type_mut`
+#[cfg(feature = "alloc")]
+pub trait Referenceable: Named {
+    /// Return the `ReReference `
+    fn type_ref(&self) -> ReReference<Self> {
```
Comment (domenukk) [inline libafl_bolts/src/tuples.rs:546]: ReReference? :D

**Pull request merged**